src/utils/async.js

```diff
@@ -1,5 +1,47 @@
 /* @flow strict-local */
-import { isClientError } from '../api/apiErrors';
+
+export class TimeoutError extends Error {
+  name = 'TimeoutError';
+}
+
+/**
+ * Time-out a Promise after `timeLimitMs` has passed.
+ *
+ * Returns a new Promise with the same outcome as `promise`, if
+ * `promise` completes in time.
+ *
+ * If `promise` does not complete before `timeLimitMs` has passed,
+ * `onTimeout` is called, and its outcome is used as the outcome of
+ * the returned Promise.
+ *
+ * If `onTimeout` is omitted, a timeout will cause the returned
+ * Promise to reject with a TimeoutError.
+ *
+ * Interface modeled after
+ * https://api.dart.dev/stable/2.8.4/dart-async/Future/timeout.html.
+ *
+ * @param promise
+ * @param timeLimitMs
+ * @param onTimeout
+ */
+export async function promiseTimeout<T>(
+  promise: Promise<T>,
+  timeLimitMs: number,
+  onTimeout?: () => T | Promise<T>,
+): Promise<T> {
+  const timeoutPromise = new Promise((_, reject) =>
+    setTimeout(() => reject(new TimeoutError()), timeLimitMs),
+  );
+  try {
+    return await Promise.race([promise, timeoutPromise]);
+  } catch (e) {
+    if (e instanceof TimeoutError && onTimeout !== undefined) {
+      return onTimeout();
```

Comment:
This feels like ideally it belongs at the same place as the `new TimeoutError()`, and replacing that in the `onTimeout` case. I think that would simplify the code, because this whole try/catch would then get to go away.

IOW, the whole try/catch could become just
```js
  return await Promise.race([promise, timeoutPromise]);
```
with a suitable version of `timeoutPromise` that itself handles the `onTimeout`.

(And then in fact the `await` could be deleted, along with the `async`: just `return Promise.race(…);`.)

Reply:
Ah, I thought of doing it that way—but then, IIUC, it's not a race between (A) the outcome of `promise` and (B) the timer (which is what it says on the tin), but rather, between (A) the outcome of `promise` and (B) the timer *plus* the outcome of a Promise returned by `onTimeout`.

Reply:
Mm, bumping this, @gnprice, as I don't think we got around to it on the call (we were mostly talking about https://github.com/zulip/zulip-mobile/pull/4166#discussion_r442571136).

Reply:
Concretely, I think you're suggesting something like this:

```javascript
  const timeoutPromise = new Promise((resolve, reject) =>
    setTimeout(
      () => (onTimeout !== undefined ? resolve(onTimeout()) : reject(new TimeoutError())),
      timeLimitMs,
    ),
  );
  return await Promise.race([promise, timeoutPromise]);
```

Reply:
But this extends the effective timeout duration by the duration of a Promise returned by `onTimeout`; I'm writing a test against this subtle case right now. 😛

Reply:
`resolve(onTimeout())` might look a bit weird if `onTimeout` returns a Promise; I'm not used to doing that. But I think this excerpt supports the idea that the outcome of `onTimeout()` (the Promise returned by `onTimeout`) is assigned to be the outcome of `timeoutPromise`.

It's from a big green note [here](https://tc39.es/ecma262/#sec-promise-executor), in a section on the Promise constructor in an ECMAScript spec (formatting lost):

> The resolve function that is passed to an executor function accepts a single argument. The executor code may eventually call the resolve function to indicate that it wishes to resolve the associated Promise object. The argument passed to the resolve function represents the eventual value of the deferred action and can be either the actual fulfillment value or another Promise object which will provide the value if it is fulfilled.

Reply:
I'll include a test for this case and you can play around with it if you have doubts. 🙂

Reply:
Hmm, good point. And good thought to add a test for it. :slightly_smiling_face: 

It still feels like this logic is sort of awkwardly split between the `new TimeoutError()` and the later `catch` and `if (e instanceof TimeoutError && onTimeout !== undefined)`. But evidently it'd take some deeper thinking about these promise and timer semantics to fix that than I'd thought. And there isn't *that* much to gain -- the behavior seems correct as it is, and the code is already short and not hard to understand.